src/lib/shell/MainLoopZephyr.cpp

```diff
@@ -14,30 +14,117 @@
 }
 
+/**
+ * This recursive function obtains all sub-commands (possbile completions) for a command node
+ * and build it into a tree of Zephyr's shell_static_entry structure.
+ *
+ * @param prefix                  The path/prefix until the root command node of the tree, without tailing space.
+ *                                e.g. "dns browse". Set to emptry string "" if building tree for a root-level command.
+ * @param syntax                  The single-word command name/syntax for the root node of the tree. e.g. "resolve".
+ *                                Set to emptry string "" if calling to build from the root (to include all commands).
+ * @param help                    The help text for the command at roote node of the tree. Set to NULL if none.
+ *
+ * @return                        The shell_static_entry structure tree that chains all the sub-commands from the
+ *                                requested command node.
+ *
+ */
+const struct shell_static_entry * build_command_tree(const char * prefix, const char * syntax, const char * help)
+{
+
+    char * sub_prefix = new char[strlen(prefix) + strlen(syntax) + 2];
+    size_t pos        = 0;
+    if (strcmp(prefix, "") != 0)
+    {
+        strncpy(&sub_prefix[pos], prefix, strlen(prefix));
+        pos += strlen(prefix);
+        strncpy(&sub_prefix[pos], " ", 1);
+        pos += 1;
+    }
+    strncpy(&sub_prefix[pos], syntax, strlen(syntax) + 1);
+
+    cmd_completion_context context = cmd_completion_context(sub_prefix);
+    CHIP_ERROR ret                 = Engine::GetCommandCompletions(&context);
+
+    const struct shell_static_entry * static_entry;
+    if (ret == CHIP_NO_ERROR && context.cmdc != 0)
+    {
+        const struct shell_static_entry * sub_static_entry = new shell_static_entry[context.cmdc + 1];
+        for (size_t i = 0; i < context.cmdc; i++)
+        {
+            shell_command_t * subcmd                    = context.cmdv[i];
+            const struct shell_static_entry * sub_entry = build_command_tree(sub_prefix, subcmd->cmd_name, subcmd->cmd_help);
+            if (sub_entry == nullptr)
+            {
+                // leaf node, the command should require no argument but can accept optional arguments
+                const struct shell_static_entry leaf_entry = {
+                    .syntax = subcmd->cmd_name, .help = subcmd->cmd_help, .subcmd = NULL, .handler = NULL, .args = { 0, 10 }
+                };
+                memcpy((shell_static_entry *) &sub_static_entry[i], &leaf_entry, sizeof(struct shell_static_entry));
+                continue;
+            }
+            memcpy((shell_static_entry *) &sub_static_entry[i], sub_entry, sizeof(struct shell_static_entry));
+        }
+
+        // Zephyr's shell_cmd_entry.u.entry needs an additional NULL element to terminate the array.
+        memset((shell_static_entry *) &sub_static_entry[context.cmdc], 0, sizeof(struct shell_static_entry));
+
+        const struct shell_cmd_entry * sub_cmd_entry =
```

Comment:
lifetime comments for these and below.
Also explanation of why this does not use PlatformAlloc but direct new